src/modules/commander/Commander.cpp

```diff
@@ -3335,177 +3356,195 @@ void *commander_low_prio_loop(void *arg)
 			switch (cmd.command) {
 			case vehicle_command_s::VEHICLE_CMD_PREFLIGHT_CALIBRATION: {
 
-					int calib_ret = PX4_ERROR;
+					if ((status.arming_state == vehicle_status_s::ARMING_STATE_ARMED)
+					    || status.arming_state == vehicle_status_s::ARMING_STATE_SHUTDOWN) {
 
-					/* try to go to INIT/PREFLIGHT arming state */
-					if (TRANSITION_DENIED == arming_state_transition(&status, safety_s{}, vehicle_status_s::ARMING_STATE_INIT, &armed,
-							false /* fRunPreArmChecks */, &mavlink_log_pub, &status_flags,
-							PreFlightCheck::arm_requirements_t{}, // arming requirements not relevant for switching to ARMING_STATE_INIT
-							30_s) // time since boot not relevant for switching to ARMING_STATE_INIT
-					   ) {
-
-						answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_DENIED, command_ack_pub);
-						break;
+						// reject if armed or shutting down
+						answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_TEMPORARILY_REJECTED, command_ack_pub);
```

Comment:
This essentially side-steps the arming "state machine" which is a bit odd, right?

Reply:
I suppose that's true, however it's not really about circumventing the arming state machine so much as skipping the entire thing when armed (or shutting down). Part of the fix is that QGC is sending `CMD_PREFLIGHT_CALIBRATION` (without any param) when you click on the RADIO tab, before any actual calibration begins.

Below this in the actual command handling where it always attempts to transition to ARMING_STATE_INIT, then checks the parameter value to perform the relevant calibration. So ideally we'd combine the command validation with the state machine transition so the command can get a proper reply. 

Overall I only saw potential downsides form `CMD_PREFLIGHT_CALIBRATION` or `CMD_PREFLIGHT_STORAGE` execution when armed (or shutting down).

Reply:
The core arming state machine is correct but if the user accidentally tries to calibrate in air and gets a message saying "ARMED to INIT denied" he doesn't even know what he did wrong. I think that's the main problem being addressed here.

Reply:
> Part of the fix is that QGC is sending CMD_PREFLIGHT_CALIBRATION (without any param) when you click on the RADIO tab

💡 That's why it does the acceptance beep and green blink when you switch to the radio tab.